### PR TITLE
state-migration: Fail if account would be overwritten

### LINKED ISSUE #173: Decide behaviour when overwriting state during migration
When a collition happens during copying the L2 contracts into the state we currently overwrite the existing state.
https://github.com/celo-org/optimism/blob/b5a2af7f414a2359a2925946c1a03f8aa88c4f9b/op-chain-ops/cmd/celo-migrate/state.go#L266

This might not be what we want to do during mainnet migration though. Alternative options include stopping the migration when this happens or logging a warning (although this risks a broken state after migration).

### PR DESCRIPTION
**Pull request by palango**
Closes https://github.com/celo-org/optimism/issues/173

### CODE REVIEW

File: op-chain-ops/cmd/celo-migrate/state.go [hunk 1]
```diff
@@ -264,9 +267,11 @@ func applyAllocsToState(db *state.StateDB, genesis *core.Genesis, config *params
 						log.Info("Account already exists with different code and is whitelisted, overwriting...", "address", k)
```
Comment (piersy) [inline op-chain-ops/cmd/celo-migrate/state.go]: Why do we have a whitelist that allows overwriting accounts?

File: op-chain-ops/cmd/celo-migrate/state.go [hunk 2]
```diff
@@ -264,9 +267,11 @@ func applyAllocsToState(db *state.StateDB, genesis *core.Genesis, config *params
...
 					} else {
 						log.Warn("Account already exists with different code and is not whitelisted, overwriting...", "address", k, "oldCode", db.GetCode(k), "newCode", v.Code)
+						return fmt.Errorf("account already exists with different code and is not whitelisted: %s", k.Hex())
```
Comment (piersy) [inline op-chain-ops/cmd/celo-migrate/state.go]: This branch seems redundant, seems to be the same as on line 274 (same account, different code, not whitelisted)

File: op-chain-ops/cmd/celo-migrate/state.go (hunk 1 above)
Comment (palango) [reply op-chain-ops/cmd/celo-migrate/state.go]: Because we know of one contract (`Create2Deployer`) which already exists but has different code on Alfajores.

File: op-chain-ops/cmd/celo-migrate/state.go (hunk 2 above)
Comment (palango) [reply op-chain-ops/cmd/celo-migrate/state.go]: I kept them separate to make the error of not supplying a whitelist more explicit.

File: op-chain-ops/cmd/celo-migrate/state.go [hunk 3]
```diff
@@ -264,9 +267,11 @@ func applyAllocsToState(db *state.StateDB, genesis *core.Genesis, config *params
...
 					}
 				} else {
 					log.Warn("Account already exists with different code and no whitelist exists", "address", k, "oldCode", db.GetCode(k), "newCode", v.Code)
+					return fmt.Errorf("account already exists with different code and no whitelist exists: %s", k.Hex())
 				}
 
 				overwriteCounter++
```
Comment (piersy) [inline op-chain-ops/cmd/celo-migrate/state.go]: Currently this function is just checking whether the code matches or not. I think I would be inclined to just fail if there is an existing account. I can't imagine a situation where we are doing this deploy and we find an account that somehow has the same code that we are already deploying.

Comment (piersy) [reply op-chain-ops/cmd/celo-migrate/state.go]: So either if there's a clash either it's whitelisted, or we fail.

Comment (palango) [reply op-chain-ops/cmd/celo-migrate/state.go]: > So either if there's a clash either it's whitelisted, or we fail.

How is that different from what the code currently does?

Comment (piersy) [reply op-chain-ops/cmd/celo-migrate/state.go]: Currently if there is a clash but the code is equal, we'll overwrite the nonce and update the balance.

Comment (palango) [reply op-chain-ops/cmd/celo-migrate/state.go]: Ok, incorporated a couple of changes:
- Check that the copied account has a balance of zero. otherwise we'd change the total supply of celo.
- Check if we'd touch an EO account with non-zero nonce, if so bail out. Thinking here is that we should not overwrite EOAs.
- If we overwrite and smart contract account, copy the nonce. Contracts have a nonce which is increased when they deploy new contracts, so we want to keep the nonce in that case.

File: op-chain-ops/cmd/celo-migrate/state.go [hunk 4]
```diff
@@ -242,40 +245,53 @@ func applyStateMigrationChanges(config *genesis.DeployConfig, genesis *core.Gene
...
 			if currentCode != nil && !equalCode {
 				if whitelist, exists := accountOverwriteWhitelist[config.ChainID.Uint64()]; exists {
 					if _, ok := whitelist[k]; ok {
-						log.Info("Account already exists with different code and is whitelisted, overwriting...", "address", k)
+						// keep the existing nonce
+						nonce = db.GetNonce(k)
+						log.Info("Account already exists with different code and is whitelisted, overwriting...", "address", k, "nonce", db.GetNonce(k))
 					} else {
-						log.Warn("Account already exists with different code and is not whitelisted, overwriting...", "address", k, "oldCode", db.GetCode(k), "newCode", v.Code)
+						log.Error("Account already exists with different code and is not whitelisted, overwriting...", "address", k, "oldCode", db.GetCode(k), "newCode", v.Code)
```
Comment (piersy) [inline op-chain-ops/cmd/celo-migrate/state.go]:
```suggestion
						log.Error("Account already exists with different code and is not whitelisted", "address", k, "oldCode", db.GetCode(k), "newCode", v.Code)
```

File: op-chain-ops/cmd/celo-migrate/state.go [hunk 5]
```diff
@@ -242,40 +245,53 @@ func applyStateMigrationChanges(config *genesis.DeployConfig, genesis *core.Gene
...
 				overwriteCounter++
 			}
 		}
+
+		// This carries over any existing balance
 		db.CreateAccount(k)
 
-		db.SetNonce(k, v.Nonce)
-		db.SetBalance(k, balance)
+		db.SetNonce(k, nonce)
```
Comment (piersy) [inline op-chain-ops/cmd/celo-migrate/state.go]: This and the SetState calls below, still don't seem correct to me.

So we could have a situation where we have an account that we are deploying that already exists with matching code and we will overwrite the nonce and update the associated state for the existing account. I don't understand why we would want to do this?

To me it seems that if we haven't whitelisted an account to overwrite, we should just fail.

Comment (palango) [reply op-chain-ops/cmd/celo-migrate/state.go]: Good point.

The logic was hard to follow, so I refactored a bit to make it more clear when accounts are copied. Let me know what you think.

File: op-chain-ops/cmd/celo-migrate/state.go [hunk 6]
```diff
@@ -242,48 +246,72 @@ func applyStateMigrationChanges(config *genesis.DeployConfig, genesis *core.Gene
...
+					return fmt.Errorf("account already exists with nonce > 0: %s", k.Hex())
+				} else {
+					copyAlloc(db, k, v, v.Nonce)
+					copyCounter++
+				}
+			} else { // account has code
+				equalCode := bytes.Equal(db.GetCode(k), v.Code)
+				if equalCode {
+					log.Info("Account already exists with same code, skipping...", "address", k.Hex())
+					// No need to overwrite the account, keep existing state and nonce
```
Comment (piersy) [inline op-chain-ops/cmd/celo-migrate/state.go]: Looking into the allocs I can see that we have accounts with nonces of 0 or 1 and also some have state.

So I don't see why it would be safe to skip updating the nonce and state in this situation. If some other part of the system is relying on the nonce and state being set to certain values then skipping here is going to break something.

Comment (palango) [reply op-chain-ops/cmd/celo-migrate/state.go]: You can also argue the other direction.

In this case there's the exact contract already deployed and we don't want to destroy any other user relying on the existing state. So we shouldn't overwrite state or nonce.

Comment (palango) [reply op-chain-ops/cmd/celo-migrate/state.go]: To conclude, I think this case is behaviour where we need to look at the particular case.

Comment (piersy) [reply op-chain-ops/cmd/celo-migrate/state.go]: I think in this case we need to fail and then manually asses the exact situation. I don't think we can safely just choose one or other approaches.

The nonce for a smart contract address is used to derive the addresses of contracts deployed by the smart contract, so having a different nonce may cause other contracts to not be able to find the contract they are looking for, or to mistakenly use a different deployment than the one they wanted or for the deployer to end up stuck unable to deploy any contracts.

Likewise I assume the state in the allocs is important, we can't just ignore it.

Comment (palango) [reply op-chain-ops/cmd/celo-migrate/state.go]: Ok, I made it fail in that case.

File: op-chain-ops/cmd/celo-migrate/state.go [hunk 7]
```diff
@@ -242,48 +246,70 @@ func applyStateMigrationChanges(config *genesis.DeployConfig, genesis *core.Gene
...
-	accountCounter := 0
+	copyCounter := 0
+	skipCounter := 0
 	overwriteCounter := 0
-	for k, v := range genesis.Alloc {
-		accountCounter++
+	whitelist := accountOverwriteWhitelist[config.ChainID.Uint64()]
 
-		balance := uint256.MustFromBig(v.Balance)
+	for k, v := range genesis.Alloc {
```
Comment (piersy) [inline op-chain-ops/cmd/celo-migrate/state.go:259]: This is getting quite complicated and I don't think it needs to be, I think the following for loop would be sufficient.

```go
	for k, v := range genesis.Alloc {
		// Check that the balance of the account to written is zero,
		// as we must not create new CELo tokens
		if v.Balance.Cmp(big.NewInt(0)) != 0 {
			log.Error("Account balance is not zero, would change celo supply", "address", k.Hex())
			return fmt.Errorf("account balance is not zero, would change celo supply: %s", k.Hex())
		}

		if db.Exist(k) {
			_, whitelisted := whitelist[k]

			// If the account is not whitelisted and has a non zero nonce or code size, bail out we will need to manually investigate how to handle this.
			if !whitelisted && (db.GetCodeSize(k) > 0 || db.GetNonce(k) > 0) {
				return fmt.Errorf("account exists and is not whitelisted, account: %s, nonce: %d, codeSize: %d", k.Hex(), db.GetNonce(k), db.GetCodeSize(k))
			}
			overwriteCounter++
		}
		copyAlloc(db, k, v, v.Nonce)
		copyCounter++
	}
```

Comment (palango) [reply op-chain-ops/cmd/celo-migrate/state.go:259]: This way we loose the specific errors, but that might be fine given that we fail anyways.

But this approach also limits our options when we would decide to keep the existing nonce in any of the cases we discussed above.

Comment (piersy) [reply op-chain-ops/cmd/celo-migrate/state.go:259]: I didn't feel that the specific errors were that helpful, they were all variations of (we're failing because there is an existing account) we could extend the error message in the code I posted above to include whether the contract code matches and then I think all the information that was included in all the specific errors should be covered.

>But this approach also limits our options when we would decide to keep the existing nonce in any of the cases we discussed above.

That is true and thinking about it more there doesn't seem to be any clear or consistent way to decide if the nonce should be kept or overwritten. I think we have to asses this on a case by case basis so the decision to keep or not keep the nonce should go into the whitelist.

In the case of the create2 deployer, it seems as if the nonce value is not particularly important, since the nonce is used to derive the contract address when calling the create method but not when calling the create2 method, and given the name of the create2 deployer it sounds like it's not going to be calling create ( I haven't actually checked). So I think we could use either. But in cases where contracts are calling create we'll need to ensure that we keep the existing nonce.

Comment (palango) [reply op-chain-ops/cmd/celo-migrate/state.go:259]: Done

Comment (piersy) [reply op-chain-ops/cmd/celo-migrate/state.go:259]: The whitelist is still lacking a decision about whether to keep the nonce. You could just make it a `map[common.Address]bool` and keep the nonce if the bool is true.

Comment (palango) [reply op-chain-ops/cmd/celo-migrate/state.go:259]: Implemented in [77fbec6](https://github.com/celo-org/optimism/pull/202/commits/77fbec69028e3b92c2b2d09356d77a8f7663bb31)

**Review by piersy — APPROVED** [review]
And it's done!
